In [2]:
#data wrangling
#data gathering --- api, csv,web scraping, databases
#assessing data-----todays topic
#cleaning data

assessing data is understending the data deeply before apply method to clean data
- so studying about unclean data and identifying lies in this accessing data catagory 

Types of Unclean Data
1. Dirty data-- also known as low quality data. means has content issue

2. messy data-- also known as untidy data. means which has structural issue

Dirty Data
- Duplicate data
- Missing data
- Corrupt Data
- Inaccurate Data

Messy data
- which have right information but unevenly divides in column and rows and you find difficulty to draw colclusions if you directly use it

* rules
1. each variable forms a column
2. each observation forms a row
3. each observation unit forms a table

Types of assessment
- Manual - looking through the data manually in google sheet
- Programmatic - by using pandas fuction such as info() describe()


Steps in Assessment

1. Discover
2. Document

In [3]:
import pandas as pd

In [4]:
patients = pd.read_csv('insulin_patient.csv')
treatment = pd.read_csv('treatment.csv')
adverse = pd.read_csv('adverse_eff.csv')

In [5]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


Quality issue

[completeness, validity, accuracy, consistency] is [cmp, v, a, con] all following quality errors lies in these catagories only

1. patients
  - zip code column is stored in decimal format and some are of 4 digit [v]
  - state coloumn some time full name and some time short form [con]
  - missing values for contact details 12[cmp] 
  - incorrect data type for sex and dob[?]
  - hight and wieght are sometimes very unnatural[a]
  - inconcistence way of storing phone no.[con]
  - dubplicate data and not usefull data[con] 

2. Treatment 
  - missing value in hba1c_change[cmp]
  - novodra and orilin column has u attached numbers[v]
  - only 280 rows are there instead of 350[cmp]
  - name and sr. name is in lowercase[con]
  - hba1c_change 9 instead 4 [a]
  - use of dashes instead of nan innovodra and insulin [con] 


3. Adverse reaction

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [9]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [11]:
treatment.describe()
# 50 to 75 is going very sharp so there is error which we saw 0.94 wala

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [13]:
patients.sample(5)

patient_id assigned_sex given_name   surname                     address  \
219         220         male         Mỹ     Quynh                         NaN   
102         103       female      Najla     Touma    2146 Willow Greene Drive   
188         189         male      Søren  Sørensen            2397 Bell Street   
165         166         male     Zlatko  Rukavina               592 Rafe Lane   
62           63       female    Firenze     Fodor  1786 Gerald L. Bates Drive   

           city state  zip_code        country  \
219         NaN   NaN       NaN            NaN   
102  Montgomery    AL   36104.0  United States   
188    New York    NY   10011.0  United States   
165  Yazoo City    MS   39194.0  United States   
62      Belmont    MA    2178.0  United States   

                                        contact   birthdate  weight  height  \
219                                         NaN    4/9/1978   237.8      69   
102  NajlaQismahTouma@cuvox.de+1 (334) 261-1235   8/20/1948   165.0      61   
188     SrenSrensen@superrito.com1 212 201 3108  12/31/1942   157.1      67   
165    ZlatkoRukavina@cuvox.de+1 (662) 716-9586  12/12/1976   198.7      70   
62          617-883-5967FodorFirenze@dayrep.com    4/1/1943   131.1      60   

      bmi  
219  35.1  
102  31.2  
188  24.6  
165  28.5  
62   25.6

In [18]:
patients[patients['address'].duplicated()]
# here jhon doe patient is repeating

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

Tidiness issues

1. patients
  - phone and email together in one contact column

2. Treatment
  - 3 variables stored in orilin and novodra column
  

3. Adverse reaction
  - no pooint using this table

Data Quality Dimensions
 1. completeness issues
 2. validity issues
 3. accuracy issues
 4. consistency issues